### ISSUE TITLE
Add for()/when() syntax in official Perl docs

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12132]
Migrated from [rt.perl.org#113064](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=113064) (status was 'resolved')

Searchable as RT113064$

**Comment by p5pRT** [comment]
### From Perl@ResonatorSoft.org

#### Created by Perl@ResonatorSoft.org


This is a bug report for perl from Perl@&#8203;ResonatorSoft\.org\,
generated with the help of perlbug 1\.39 running under perl 5\.10\.1\.

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
From IRC&#8203;:

Original topic&#8203;: http&#8203;://www\.effectiveperlprogramming\.com/blog/1333

Should we put this into the official Perl docs?  Otherwise\, I
imagine the lifecycle of a programmer's use of given/when will be
this&#8203;:

1\. "Hey\, this is great\!  Finally\, a switch/case type syntax\."
2\. "WTF?  My any \{ $\_ stuff \} doesn't work as expected\."
3\. "Wait\, this is SUPPOSED to work like this?"
4\. "Okay\, I'm swearing off given/when\."
5\. "Oh wait\, so you can actually still use it without all of the
&nbsp;    $\_ problems?  Why isn't this in the Perl docs to begin with?"

With various programmers in various states of 1 thru 5\.

<details><summary>Perl Info</summary>

```

Flags:
    category=docs
    severity=medium

Site configuration information for perl 5.10.1:

Configured by Debian Project at Thu Jun 30 19:24:11 UTC 2011.

Summary of my perl5 (revision 5 version 10 subversion 1) configuration:

TL;DR

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Wed May 23 15&#8203;:51&#8203;:31 2012\, Perl@&#8203;ResonatorSoft\.org wrote&#8203;:
> 
> This is a bug report for perl from Perl@&#8203;ResonatorSoft\.org\,
> generated with the help of perlbug 1\.39 running under perl 5\.10\.1\.
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> From IRC&#8203;:
> 
> Original topic&#8203;: http&#8203;://www\.effectiveperlprogramming\.com/blog/1333
> 
> Should we put this into the official Perl docs?  Otherwise\, I
> imagine the lifecycle of a programmer's use of given/when will be
> this&#8203;:
> 
> 1\. "Hey\, this is great\!  Finally\, a switch/case type syntax\."
> 2\. "WTF?  My any \{ $\_ stuff \} doesn't work as expected\."
> 3\. "Wait\, this is SUPPOSED to work like this?"
> 4\. "Okay\, I'm swearing off given/when\."
> 5\. "Oh wait\, so you can actually still use it without all of the
>     $\_ problems?  Why isn't this in the Perl docs to begin with?"
> 
> With various programmers in various states of 1 thru 5\.
> 

Given the argument that brian d foy makes on that blog post\, it seems to
me that accepting his critique would more logically lead to eliminating
given/when entirely\.

\(Because almost all my code has had to be written to be 5\.8\-compatible\,
I have yet to use given/when myself\.  So I have no dog in this fight\.\)

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed May 23 15&#8203;:51&#8203;:31 2012\, Perl@&#8203;ResonatorSoft\.org wrote&#8203;:
> 
> This is a bug report for perl from Perl@&#8203;ResonatorSoft\.org\,
> generated with the help of perlbug 1\.39 running under perl 5\.10\.1\.
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> From IRC&#8203;:
> 
> Original topic&#8203;: http&#8203;://www\.effectiveperlprogramming\.com/blog/1333
> 
> Should we put this into the official Perl docs?  Otherwise\, I
> imagine the lifecycle of a programmer's use of given/when will be
> this&#8203;:
> 
> 1\. "Hey\, this is great\!  Finally\, a switch/case type syntax\."
> 2\. "WTF?  My any \{ $\_ stuff \} doesn't work as expected\."
> 3\. "Wait\, this is SUPPOSED to work like this?"
> 4\. "Okay\, I'm swearing off given/when\."
> 5\. "Oh wait\, so you can actually still use it without all of the
>     $\_ problems?  Why isn't this in the Perl docs to begin with?"

But if we suggest for/when\, people will still run into when’s bugs\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed May 23 15&#8203;:51&#8203;:31 2012\, Perl@&#8203;ResonatorSoft\.org wrote&#8203;:
> 
> This is a bug report for perl from Perl@&#8203;ResonatorSoft\.org\,
> generated with the help of perlbug 1\.39 running under perl 5\.10\.1\.
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> From IRC&#8203;:
> 
> Original topic&#8203;: http&#8203;://www\.effectiveperlprogramming\.com/blog/1333
> 
> Should we put this into the official Perl docs?  Otherwise\, I
> imagine the lifecycle of a programmer's use of given/when will be
> this&#8203;:
> 
> 1\. "Hey\, this is great\!  Finally\, a switch/case type syntax\."
> 2\. "WTF?  My any \{ $\_ stuff \} doesn't work as expected\."
> 3\. "Wait\, this is SUPPOSED to work like this?"
> 4\. "Okay\, I'm swearing off given/when\."
> 5\. "Oh wait\, so you can actually still use it without all of the
>     $\_ problems?  Why isn't this in the Perl docs to begin with?"
> 
> With various programmers in various states of 1 thru 5\.

In bleadperl\, given now aliases $\_ exactly the way for does\.  So I think
we can consider this resolved\.

Please holler if you disagree\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @epa

There must now be a fair bit of code using for/when\.  It would be worth
at least a brief mention in the documentation\.

\-\- 
Ed Avis \<eda@&#8203;waniasset\.com>

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Sep 20 05&#8203;:49&#8203;:03 2012\, eda@&#8203;waniasset\.com wrote&#8203;:
> There must now be a fair bit of code using for/when\.  It would be worth
> at least a brief mention in the documentation\.
> 

perlsyn mentions it&#8203;:

The C\<when> modifier is an experimental feature that first appeared in Perl
5\.14\.  To use it\, you should include a C\<use v5\.14> declaration\.
\(Technically\, it requires only the C\<switch> feature\, but that aspect of it
was not available before 5\.14\.\)  Operative only from within a C\<foreach>
loop or a C\<given> block\, it executes the statement only if the smartmatch
C\<\< $\_ ~~ I\<EXPR> >> is true\.  If the statement executes\, it is followed by
a C\<next> from inside a C\<foreach> and C\<break> from inside a C\<given>\.


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From bmb@Mail.Libs.UGA.EDU

On 9/20/12\, Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>
> perlsyn mentions it&#8203;:
>
> The C\<when> modifier is an experimental feature that first appeared in Perl
> 5\.14\.  To use it\, you should include a C\<use v5\.14> declaration\.
> \(Technically\, it requires only the C\<switch> feature\, but that aspect of it
> was not available before 5\.14\.\)  Operative only from within a C\<foreach>
> loop or a C\<given> block\, it executes the statement only if the smartmatch
> C\<\< $\_ ~~ I\<EXPR> >> is true\.  If the statement executes\, it is followed by
> a C\<next> from inside a C\<foreach> and C\<break> from inside a C\<given>\.

Except that it isn't followed by a simple C\<next> from inside a
C\<foreach> as you showed in the exchange below \(from another thread\)\.
It sounds like that's not likely to change\.

Brad

On 9/2/12\, Ricardo Signes \<perl\.p5p@&#8203;rjbs\.manxome\.org> wrote&#8203;:
> \* Father Chrysostomos \<sprout@&#8203;cpan\.org> \[2012\-09\-02T19&#8203;:06&#8203;:43\]
>> And here is another aspect of when/default that makes the way it breaks
>> special&#8203;:
>>
>> for \(3\) \{
>>   for my $x \(4\) \{
>>     when \(3\) \{\}
>>   \}
>>   warn "exited inner";
>> \}
>> warn "exited outer";
>>
>> That prints "exited outer at \- line 7"\.  So when doesn’t just do ‘next’
>> inside for\.
>>
>> I think that should change\.
>
> Well\, the issue is that when interacts with a topicalizer\.  It is going to
> do
> its work based on the topicalization that has occurred\, and so having its
> implicit break behave like a 'next' on the relevant topicalizer is not
> obviously nuts\.

**Comment by p5pRT** [comment]
### From @doy

On Thu\, Sep 20\, 2012 at 10&#8203;:32&#8203;:54AM \-0400\, Brad Baxter wrote&#8203;:
> On 9/20/12\, Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> >
> > perlsyn mentions it&#8203;:
> >
> > The C\<when> modifier is an experimental feature that first appeared in Perl
> > 5\.14\.  To use it\, you should include a C\<use v5\.14> declaration\.
> > \(Technically\, it requires only the C\<switch> feature\, but that aspect of it
> > was not available before 5\.14\.\)  Operative only from within a C\<foreach>
> > loop or a C\<given> block\, it executes the statement only if the smartmatch
> > C\<\< $\_ ~~ I\<EXPR> >> is true\.  If the statement executes\, it is followed by
> > a C\<next> from inside a C\<foreach> and C\<break> from inside a C\<given>\.
> 
> Except that it isn't followed by a simple C\<next> from inside a
> C\<foreach> as you showed in the exchange below \(from another thread\)\.
> It sounds like that's not likely to change\.

From what I understand\, 4ccab1925a7 \(part of the smartmatch/given/when
reimplementation that Father C is working on\) does turn it into a simple
'next'\. Not entirely sure about the consensus for that decision\, but I
don't think I'd be against it\.

\-doy

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Sep 20 09&#8203;:30&#8203;:34 2012\, doy@&#8203;tozt\.net wrote&#8203;:
> On Thu\, Sep 20\, 2012 at 10&#8203;:32&#8203;:54AM \-0400\, Brad Baxter wrote&#8203;:
> > On 9/20/12\, Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org>
> wrote&#8203;:
> > >
> > > perlsyn mentions it&#8203;:
> > >
> > > The C\<when> modifier is an experimental feature that first
> appeared in Perl
> > > 5\.14\.  To use it\, you should include a C\<use v5\.14> declaration\.
> > > \(Technically\, it requires only the C\<switch> feature\, but that
> aspect of it
> > > was not available before 5\.14\.\)  Operative only from within a
> C\<foreach>
> > > loop or a C\<given> block\, it executes the statement only if the
> smartmatch
> > > C\<\< $\_ ~~ I\<EXPR> >> is true\.  If the statement executes\, it is
> followed by
> > > a C\<next> from inside a C\<foreach> and C\<break> from inside a
> C\<given>\.
> >
> > Except that it isn't followed by a simple C\<next> from inside a
> > C\<foreach> as you showed in the exchange below \(from another
> thread\)\.
> > It sounds like that's not likely to change\.
> 
> From what I understand\, 4ccab1925a7 \(part of the smartmatch/given/when
> reimplementation that Father C is working on\)

At this point I’m not working on it\.  I’m just waiting for feedback from
rjbs\.

> does turn it into a
> simple
> 'next'\. Not entirely sure about the consensus for that decision\, but I
> don't think I'd be against it\.


\-\- 

Father Chrysostomos
